File: BAEKJOON/1446 - 지름길/지름길.cpp
#include <iostream>  // 입출력 처리
#include <vector>    // 벡터 자료형 사용
#include <queue>     // 우선순위 큐 사용
#include <algorithm> // min 함수 사용

using namespace std;

struct Shortcut {
    int start, end, length;  // 지름길의 시작점, 끝점, 지름길 길이
};

int main() {
    int N, D;  // N은 지름길의 개수, D는 고속도로의 길이
    cin >> N >> D;  // N과 D 입력받기

    vector<Shortcut> shortcuts(N);  // N개의 지름길 정보를 저장할 벡터
    for (int i = 0; i < N; ++i) {
        cin >> shortcuts[i].start >> shortcuts[i].end >> shortcuts[i].length;  // 각 지름길의 시작점, 끝점, 길이 입력받기
    }

    vector<int> dist(D + 1, D);  // 거리 배열을 고속도로 길이 D로 초기화 (최댓값은 직선으로 이동한 거리)
    dist[0] = 0;  // 시작점 0의 거리는 0으로 설정

    for (int i = 0; i <= D; ++i) {
        // 한 칸씩 직선 이동하는 경우를 먼저 처리
        if (i > 0) {
            dist[i] = min(dist[i], dist[i - 1] + 1);  // 직선으로 한 칸씩 이동하는 최소 거리 갱신
        }

        // 지름길을 사용하는 경우 처리
        for (const auto& shortcut : shortcuts) {
            if (shortcut.start == i && shortcut.end <= D) {
                dist[shortcut.end] = min(dist[shortcut.end], dist[i] + shortcut.length);  // 지름길을 이용한 이동 거리 갱신
            }
        }
    }

    cout << dist[D] << endl;  // 최종적으로 D 지점까지의 최소 거리를 출력
    return 0;
}

File: BAEKJOON/1446 - 지름길/지름길.cs
using System;
using System.Collections.Generic;

class Shortcut
{
    public int Start { get; set; }
    public int End { get; set; }
    public int Length { get; set; }
}

class Program
{
    static void Main(string[] args)
    {
        int N, D;  // N은 지름길의 개수, D는 고속도로의 길이
        string[] input = Console.ReadLine().Split();  // 첫 번째 줄 입력받기
        N = int.Parse(input[0]);  // N값 입력
        D = int.Parse(input[1]);  // D값 입력

        List<Shortcut> shortcuts = new List<Shortcut>();  // 지름길 리스트 생성

        for (int i = 0; i < N; i++)
        {
            input = Console.ReadLine().Split();  // 지름길 정보 입력
            int start = int.Parse(input[0]);
            int end = int.Parse(input[1]);
            int length = int.Parse(input[2]);
            shortcuts.Add(new Shortcut { Start = start, End = end, Length = length });  // 각 지름길 정보를 리스트에 추가
        }

        int[] dist = new int[D + 1];  // 각 지점까지의 최소 거리를 저장할 배열
        for (int i = 0; i <= D; i++)
            dist[i] = i;  // 처음에는 직선으로 이동한 거리를 초기화

        for (int i = 0; i <= D; i++)
        {
            if (i > 0)
            {
                dist[i] = Math.Min(dist[i], dist[i - 1] + 1);  // 직선 이동 경로 처리
            }

            foreach (var shortcut in shortcuts)
            {
                if (shortcut.Start == i && shortcut.End <= D)
                {
                    dist[shortcut.End] = Math.Min(dist[shortcut.End], dist[i] + shortcut.Length);  // 지름길을 이용한 경로 갱신
                }
            }
        }

        Console.WriteLine(dist[D]);  // 최종 결과 출력
    }
}
